File: 2213-longest-substring-of-one-repeating-character/2213-longest-substring-of-one-repeating-character.py
from sortedcontainers import SortedList

class Solution:
    def longestRepeating(self, s: str, queryCharacters: str, queryIndices: List[int]) -> List[int]:
        N = len(s)
        mp = defaultdict(SortedList)
        i = 0
        res = []
        B = SortedList()
        while i < N:
            j = i+1
            while j < N and s[j] == s[i]:
                j += 1
            mp[s[i]].add([i,j])
            B.add(j-i)
            i = j
        s = [ch for ch in s]
        for ch,i in zip(queryCharacters,queryIndices):
            o_ch = s[i]
            if o_ch == ch:
                res.append(B[-1])
                continue
            s[i] = ch
            A = mp[ch]
            if len(A) == 0:
                A.add([i,i+1])
                B.add(1)
            else:
                p = A.bisect_right([i,inf])
                if p == 0:
                    if A[p][0] == i+1:
                        B.discard(A[p][1] - A[p][0])
                        A[p][0] = i
                        B.add(A[p][1] - A[p][0])
                    else:
                        A.add([i,i+1])
                        B.add(1)
                elif p == len(A):
                    if A[p-1][1] == i:
                        B.discard(A[p-1][1] - A[p-1][0])
                        A[p-1][1] = i+1
                        B.add(A[p-1][1] - A[p-1][0])
                    else:
                        A.add([i,i+1])
                        B.add(1)
                else:
                    if A[p-1][1] < i and A[p][0] > i+1:
                        A.add([i,i+1])
                        B.add(1)
                    elif A[p-1][1] == i and A[p][0] == i+1:
                        B.discard(A[p][1] - A[p][0])
                        B.discard(A[p-1][1] - A[p-1][0])
                        A[p-1][1] = A[p][1]
                        A.discard(A[p])
                        B.add(A[p-1][1] - A[p-1][0])
                    elif A[p-1][1] == i:
                        B.discard(A[p-1][1] - A[p-1][0])
                        A[p-1][1] = i+1
                        B.add(A[p-1][1] - A[p-1][0])
                    else:
                        B.discard(A[p][1] - A[p][0])
                        A[p][0] = i
                        B.add(A[p][1] - A[p][0])
            # print(ch,A)
            
            A = mp[o_ch]
            p = A.bisect_right([i,inf])
            p -= 1
            l,r = A[p]
            B.discard(r-l)
            if r - l == 1:
                A.discard(A[p])
            elif l == i:
                A[p][0] = i+1
                B.add(r-l-1)
            elif A[p][1] == i+1:
                A[p][1] = i
                B.add(r-l-1)
            else:
                A[p][1] = i
                A.add([i+1,r])
                B.add(i-l)
                B.add(r-(i+1))
            # print(o_ch,A)
            # print(B)
            
            res.append(B[-1])
            
        return res
        

File: 2213-longest-substring-of-one-repeating-character/2213-longest-substring-of-one-repeating-character.cpp
class Solution {
public:
    struct node {
        char lc = 0, rc = 0;
        int pre = 1, suf = 1, longest = 1, sz = 1;
        void merge(node& l, node& r) {
            longest = max(l.longest, r.longest);
            if (l.rc == r.lc) {
                longest = max(longest, l.suf + r.pre);
            }
            lc = l.lc;
            rc = r.rc;
            sz = l.sz + r.sz;
            pre = l.pre + (l.rc == r.lc && l.pre == l.sz? r.pre : 0);
            suf = r.suf + (l.rc == r.lc && r.suf == r.sz? l.suf : 0);
        }
    };
    
    void update(vector<node>& T, int i, int l, int r, int p, char ch) {
        if (l <= p && p < r) {
            if (l == r-1) {
                T[i].lc = T[i].rc = ch;
            }
            else {
                int mi = (l+r)/2;
                update(T, i*2+1, l, mi, p, ch);
                update(T, i*2+2, mi, r, p, ch);
                T[i].merge(T[i*2+1], T[i*2+2]);
            }
        }
    }
    
    int query(vector<node>& T, int i, int l, int r, int ql, int qr) {
        int res = 0;
        if (l <= ql && qr <= r) {
            res = T[i].longest;
        }
        else if (r <= ql || qr <= l) {
            return 0;
        }
        else {
            int mi = (l+r)/2;
            res = max(query(T, i*2+1, l, mi, ql, qr), query(T, i*2+2, mi, r, ql, qr));
        }
        return res;
    }
    
    vector<int> longestRepeating(string s, string queryCharacters, vector<int>& queryIndices) {
        vector<int> res;
        int powOf2 = 1, sz = s.size();
        while (powOf2 < sz) 
            powOf2 <<= 1;
        vector<node> T(powOf2*2);
        for (int i = 0; i < sz; i++) {
            update(T, 0, 0, sz, i, s[i]);
        }
        for (int i = 0; i < queryCharacters.size(); i++) {
            update(T, 0, 0, sz, queryIndices[i], queryCharacters[i]);
            res.push_back(query(T, 0, 0, sz, 0, sz));
        }
        return res;
    }
};